### ISSUE TITLE
Miscompile on x64 with certain `-march` flags?

### ISSUE THREAD

**Issue by carlocab** [LuaJIT/LuaJIT#865]
When building LuaJIT (from commit 50936d784474747b4569d988767f1b5bab8bb6d0) with either no specification for `CCOPT_x64` or `CCOPT_x64=-march=syklake`, we have:
```
$ touch empty.lua; ./luajit -b -o osx -a arm empty.lua empty.o; file empty.o; rm empty.o
empty.o: Mach-O universal binary with 2 architectures: [i386: Mach-O i386 object] [arm_v7: Mach-O arm_v7 object] [architecture=16444926]
```

However, when building with `CCOPT_x64=-march=knl`, `CCOPT_x64=-march=icelake-client`, or `CCOPT_x64=-march=skylake-avx512`,
```
$ touch empty.lua; ./luajit -b -o osx -a arm empty.lua empty.o; file empty.o
empty.o: DOS executable (block device driver)
```

This occurs on both macOS and Linux, with both `gcc-11` and `clang`.

Not really sure what's going on here, but I'm happy to provide more debug information if required.

**Comment by MikePall** [comment]
I don't have a Skylake or Icelake CPU to test this on. But `-march=knl` with GCC 11.2.1_git20220219 on Linux runs fine and produces a correct object file. Ditto for `-march=native` or whatever I throw at it, as long as the CPU can still run the resulting binary. I've tried on a Core and a Xeon (not the newest generation).

The question is whether the LuaJIT executable is faulty in general (fails most of the test suite) or only shows the particular failure with `-b`.

[Note that a GC64 mode x64 executable will not generate proper bytecode for (non-GC64) 32 bit ARM. But that's tangential to the issue here.]

**Comment by carlocab** [comment]
> I don't have a Skylake or Icelake CPU to test this on. But `-march=knl` with GCC 11.2.1_git20220219 on Linux runs fine and produces a correct object file. Ditto for `-march=native` or whatever I throw at it, as long as the CPU can still run the resulting binary. I've tried on a Core and a Xeon (not the newest generation).

Hm, ok. Since it's not clear how reproducible this is, we can (ab)use GitHub infra to test on a range of OSs and `-march` flags. The results of this can be found [here](https://github.com/carlocab/LuaJIT/actions/runs/2667712658).

I built LuaJIT with a variety of `-march` flags:
```
arch: [core2, nehalem, westmere, sandybridge, ivybridge, haswell, broadwell, skylake, knl, skylake-avx512, native]
```
This was done on each of the following OSs:
```
os: [ubuntu-18.04, ubuntu-20.04, ubuntu-22.04, macOS-10.15, macOS-11, macOS-12]
```

For most of these cases, doing `./luajit -b` worked properly whenever the build `luajit` was executable. It failed to produce the right magic numbers on the following OS-march combinations:
- `ubuntu-20.04` with `knl`, `skylake-avx512`, `native`
- `ubuntu-22.04` with `knl`, `skylake-avx512`, `native`

Of course, this could be due to differing toolchain versions:
- `ubuntu-18.04` has `gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0`
- `ubuntu-20.04` has `gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0`
- `ubuntu-22.04` has `gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0`

Or, indeed, this could still be due to any number of things. But at least it does demonstrate that it's not just my machine.

The macOS results aren't really indicative since they all failed to execute `luajit` with `-march=broadwell` or newer.

> The question is whether the LuaJIT executable is faulty in general (fails most of the test suite) or only shows the particular failure with `-b`.

How do you suggest determining this?

There are tests available at https://github.com/LuaJIT/LuaJIT-test-cleanup, but that repo hasn't been touched in years. There's also https://github.com/openresty/luajit2-test-suite, but that seems to be aimed at testing their fork of LuaJIT. I'll try these tests out anyway to see if they tell me anything.

> [Note that a GC64 mode x64 executable will not generate proper bytecode for (non-GC64) 32 bit ARM. But that's tangential to the issue here.]

Noted, thanks. This issue is also reproducible with `-a arm64`.

**Comment by MikePall** [comment]
I'm beginning to suspect these compiler flags cause a change in the ABI. In particular which registers or flags are saved/restored. This could of course break the LuaJIT VM, if the assembler part is not updated accordingly.

The one thing in common is they enable AVX512. Upon a quick look in the ABI docs, I found nothing relevant. I'm too time-constrained to dive into the GCC or LLVM sources right now.

**Comment by MikePall** [comment]
I found the cause of this after going down a very deep rabbit hole. My intuition it has to do with AVX512 was correct.

AVX512F has the `vcvttsd2usi` instruction which converts double/float -> uint32_t/uint64_t. Earlier architectures (SSE2, AVX2) are sorely lacking such an instruction, as they only support signed conversions. Unsigned conversions are done with a signed convert and range shifting -- the exact algorithm depends on the compiler. A side-effect of these workarounds is that negative doubles/floats often inadvertently convert 'as expected', even though this is invoking undefined behavior. Whereas `vcvttsd2usi` always returns 0x80000000 or 0x8000000000000000 for out-of-range inputs.

In the context of #415 I tried to fix this partially by hand-coding the double -> uint64_t case for certain architectures. But the double -> uint32_t case remained untouched.

It just so happens that my own FFI code for Mach-O file generation in `bcsave.lua` relied on undefined behavior for conversions to uint32_t. That failed when LuaJIT was compiled for an architecture with AVX512F enabled.

I've fixed the code in `bcsave.lua`.

However the real issue remains unfixed. It's one of the last big things to fix in v2.1 on my list. Programmer expectation trumps implementation efficiency, so the plan is to make the behavior defined. The double to uint64_t conversions is then defined for the (asymmetric) range `-2^63 <= d < 2^64`. Similarly for uint32_t.

But don't hold your breath, I need to do this for all architectures consistently for the interpreter and the JIT compiler backend. And in assembler, not C -- see above.
